content/docs/rfcs/32/README.md

```diff
@@ -0,0 +1,554 @@
+the verification continues, otherwise the message is discarded.
+
+**4. spam verification**
+
+After the proof is verified The `x`, and `y` fields are added to the `x_shares` and `y_shares` arrays of the `messaging_metadata` `external_nullifier` and `internal_nullifier` object. 
+If the length of the arrays is equal to the spam threshold (`limit`), the user can be slashed.
+
+#### Slashing
+
+After the verification, the user can be slashed if enough shares are present to reconstruct their `identity_secret_hash` from `x_shares` and `y_shares` fields, 
+for their `internal_nullifier`.
+The secret can be retreived by the properties of the Shamir's secret sharing scheme. 
+In particular the secret (`a_0`) can be retrieved by computing `Lagrange polynomials`. 
+
+After the secret is retreived,
+ the user's `identity_commitment` can be generated from the secret and it can be used for removing the user from the membership merkle tree (zeroing out the leaf that contains the user's `identity_commitment`). 
+Additionally, depending on the application the `identity_secret_hash` can be used for taking the user's provided stake.
+
+# Technical overview
+
+The main RLN construct is implemented using a [ZK-SNARK](https://z.cash/technology/zksnarks/) circuit. 
+However it is helpful to describe the other necessary outside components for interaction with the circuit, 
+which together with the ZK-SNARK circuit enable the above mentioned features.
+
+
+## Terminology
+
+| Term                      | Description                                                                         |
+|---------------------------|-------------------------------------------------------------------------------------|
+| **ZK-SNARK**    | https://z.cash/technology/zksnarks/              |
+| **Stake**                 | Financial or social stake required for registering in the RLN applications.  Common stake examples are: locking cryptocurrency (financial), linking reputable social identity. |
+| **Identity secret**       | An array of two unique random components (identity nullifier and identity trapdoor), which must be kept private by the user. Secret hash and identity commitment are derived from this array.        |
+| **Identity nullifier**    | Random 32 byte value used as component for identity secret generation.              |
+| **Identity trapdoor**     | Random 32 byte value used as component for identity secret generation.              |
+| **Identity secret hash**  | The hash of the identity secret, obtained using the Poseidon hash function. It is used for deriving the identity commitment of the user, and as a private input for zk proof generation. The secret hash should be kept private by the user.                            |
+| **Identity commitment**   | Hash obtained from the `Identity secret hash` by using the poseidon hash function. It is used by the users for registering in the protocol.                                       |
```

Comment:
This is a great question. The hashing scheme used here is based on the design decisions which also include the Semaphore circuits. Our goal was to ensure compatibility of the secrets for apps that use Semaphore and RLN circuits while also not compromising on security because of using the same secrets.

For example let's say there is a voting app that uses Semaphore, and also a chat app that uses RLN. The UX would be better if the users would not need to care about complicated identity management (secrets and commitments) they use for each app, and it would be much better if they could use a single id commitment for this. Also in some cases these kind of dependency is required - RLN chat app using Interep as a registry (instead of using financial stake). 
One potential concern about this interoperability is a slashed user on the RLN app side having their security compromised on the semaphore side apps as well. I.e obtaining the user's secret, anyone would be able to generate valid semaphore proofs as the slashed user. We don't want that, and we should keep user's app specific security threats in the domain of that app alone.

To achieve the above interoperability UX while preventing the shared app security model (i.e slashing user on an RLN app having impact on Semaphore apps), we had to do the follow in regard the identity secret and identity commitment:

Identity secret = [nullifier, trapdoor]
Identity secret hash = poseidon(Identity secret)
Identity commitment = poseidon(Identity secret hash)

Secret components for generating Semaphore proof:
- nullifier
- trapdoor

Secret components for generting RLN proof:
- identity secret hash

When a user is slashed on the RLN app side, their identity secret hash is revealed. However a semaphore proof can't be generated because we do not know the user's nullifier and trapdoor.
With this design we achieve:

id commitment (Semaphore) == id commitment (RLN)
secret (semaphore) != secret (RLN).

This is the only option we had for the scheme in order to satisfy the properties described above.

Also for RLN we do a single secret component input for the circuit. Thus we need to hash the secret array (two components) to a secret hash, and we use that as a secret component input.

Reply:
Thanks for the very clear explanation.  This is an important design choice and is worth being included at the end of the spec in a separate section like `#Security consideration`  or `#compatibility with Semaphore`.

Reply:
Added in a new appendix:
Appendix B: Identity scheme choice